File: lexer/lexer.go
package lexer

import "strconv"

type Lexer struct {
	source   string
	pos      int
	peek_pos int
	line     int
	column   int
	cur_byte byte
}

func New(src string) *Lexer {
	l := &Lexer{
		source:   src,
		pos:      0,
		peek_pos: 0,
		line:     1,
		column:   0,
		cur_byte: 0,
	}
	l.next_char()
	return l
}

func (l *Lexer) is_at_end() bool {
	return l.pos >= len(l.source)
}

func (l *Lexer) advance_line() {
	l.line += 1
	l.column = 0
}

func (l *Lexer) next_char() {
	if l.peek_pos >= len(l.source) {
		l.cur_byte = 0
	} else {
		l.cur_byte = l.source[l.peek_pos]
	}
	l.pos = l.peek_pos
	l.peek_pos += 1
	l.column += 1
}

func (l *Lexer) peek_char() byte {
	if l.peek_pos >= len(l.source) {
		return 0
	} else {
		return l.source[l.peek_pos]
	}
}

func (l *Lexer) skip_comments_and_whitespace() {
	for !l.is_at_end() {
		if l.cur_byte == ' ' ||
			l.cur_byte == '\t' ||
			l.cur_byte == '\n' ||
			l.cur_byte == '\r' {
			if l.cur_byte == '\n' {
				l.advance_line()
			}
			l.next_char()
		} else if l.cur_byte == '/' { // if we have seen a /
			if l.peek_char() == '/' { // and seen another /, then single line comment
				l.next_char() // move to the second /
				for !l.is_at_end() && l.cur_byte != '\n' {
					l.next_char()
				}
				l.advance_line()
				l.next_char()
			} else if l.peek_char() == '*' { // multi line
				l.next_char() // move to the *
				l.next_char() // move past the *
				for !l.is_at_end() {
					if l.cur_byte == '*' {
						// we might be potentially at the end of the comment
						if l.peek_char() == '/' {
							// we are at the end of the comment
							l.next_char() // advance to the / character
							l.next_char() // advance past the / character
							break
						} else { // we are not at the end of the loop so move past the *
							l.next_char()
						}
					} else if l.cur_byte == '\n' { // register a newline and keep moving
						l.advance_line()
						l.next_char()
					} else { // skip past comment content
						l.next_char()
					}
				}
			} else {
				return
			}
		} else { // we are not at a comment so return
			return
		}
	}
}

func (l *Lexer) Next_Token() ProtoToken {
	l.skip_comments_and_whitespace()
	var token ProtoToken
	if l.is_at_end() {
		token = l.make_token(END, "EOF")
		token.TokenSpan = Span{
			Line:  l.line,
			Col:   l.column,
			Start: l.pos,
			End:   l.pos,
		}
		return token
	}

	cur_char := l.cur_byte
	switch cur_char {
	case '=':
		if l.peek_char() == '=' {
			l.next_char() // go to second =
			l.next_char() // skip past second =
			token = l.make_token(IS_EQUAL_TO, "==")
		} else {
			l.next_char() // skip past the =
			token = l.make_singlechar_token(ASSIGN, cur_char)
		}
	case '!':
		if l.peek_char() == '=' {
			l.next_char()
			l.next_char()
			token = l.make_token(NOT_EQUAL_TO, "!=")
		} else { // erroneous
			l.next_char()
			token = l.make_singlechar_token(ERROR, cur_char)
		}
	case '+':
		if l.peek_char() == '=' {
			l.next_char()
			l.next_char()
			token = l.make_token(PLUS_EQUAL, "+=")
		} else {
			l.next_char()
			token = l.make_singlechar_token(PLUS, cur_char)
		}
	case '-': // TODO: allow negative number creation
		if l.peek_char() == '=' {
			l.next_char()
			l.next_char()
			token = l.make_token(MINUS_EQUAL, "-=")
		} else if l.peek_char() == '>' {
			l.next_char()
			l.next_char()
			token = l.make_token(ARROW, "->")
		} else {
			l.next_char()
			token = l.make_singlechar_token(MINUS, cur_char)
		}
	case '*':
		if l.peek_char() == '=' {
			l.next_char()
			l.next_char()
			token = l.make_token(STAR_EQUAL, "*=")
		} else {
			l.next_char()
			token = l.make_singlechar_token(STAR, cur_char)
		}
	case '/':
		if l.peek_char() == '=' {
			l.next_char()
			l.next_char()
			token = l.make_token(SLASH_EQUAL, "/=")
		} else {
			l.next_char()
			token = l.make_singlechar_token(SLASH, cur_char)
		}
	case '%':
		if l.peek_char() == '=' {
			l.next_char()
			l.next_char()
			token = l.make_token(MODULO_EQUAL, "%=")
		} else {
			l.next_char()
			token = l.make_singlechar_token(MODULO, cur_char)
		}
	case '\'': // read character and handle errors
		token = l.read_character()
	case '"': // read string and handle errors
		token = l.read_string()
	case '<':
		if l.peek_char() == '=' {
			l.next_char() // go to =
			l.next_char() // skip past =
			token = l.make_token(LESS_OR_EQUAL, "<=")
		} else {
			l.next_char() // skip past the =
			token = l.make_singlechar_token(LESS_THAN, cur_char)
		}
	case '>':
		if l.peek_char() == '=' {
			l.next_char() // go to =
			l.next_char() // skip past =
			token = l.make_token(GREATER_OR_EQUAL, ">=")
		} else {
			l.next_char() // skip past the =
			token = l.make_singlechar_token(GREATER_THAN, cur_char)
		}
	case '&':
		if l.peek_char() == '&' {
			l.next_char()
			l.next_char()
			token = l.make_token(AND, "&&")
		} else { // erroneous
			l.next_char() // skip past |
			token = l.make_singlechar_token(ERROR, cur_char)
		}
	case '|':
		if l.peek_char() == '|' {
			l.next_char()
			l.next_char()
			token = l.make_token(OR, "||")
		} else {
			l.next_char() // skip past |
			token = l.make_singlechar_token(PIPE, cur_char)
		}
	case ',':
		l.next_char()
		token = l.make_singlechar_token(COMMA, cur_char)
	case '.':
		if l.peek_char() == '.' {
			l.next_char()
			l.next_char()

			if l.cur_byte == '=' {
				l.next_char()
				token = ProtoToken{
					Type:    INCLUSIVE_RANGE,
					Literal: "..=",
					TokenSpan: Span{
						Line:  l.line,
						Col:   l.column - 3,
						Start: l.pos - 3,
						End:   l.peek_pos,
					},
				}
			} else {
				token = l.make_token(RANGE, "..")
			}
		} else {
			l.next_char()
			token = l.make_singlechar_token(DOT, cur_char)
		}
	case ';':
		l.next_char()
		token = l.make_singlechar_token(SEMI_COLON, cur_char)
	case ':':
		l.next_char()
		token = l.make_singlechar_token(COLON, cur_char)
	case '?':
		l.next_char()
		token = l.make_singlechar_token(QUESTION_MARK, cur_char)
	case '(':
		l.next_char()
		token = l.make_singlechar_token(OPEN_PAREN, cur_char)
	case ')':
		l.next_char()
		token = l.make_singlechar_token(CLOSE_PAREN, cur_char)
	case '[':
		l.next_char()
		token = l.make_singlechar_token(OPEN_BRACKET, cur_char)
	case ']':
		l.next_char()
		token = l.make_singlechar_token(CLOSE_BRACKET, cur_char)
	case '{':
		l.next_char()
		token = l.make_singlechar_token(OPEN_CURLY, cur_char)
	case '}':
		l.next_char()
		token = l.make_singlechar_token(CLOSE_CURLY, cur_char)
	default: // handle digits, keywords and identifiers
		if is_alphabet(cur_char) || cur_char == '_' { // check for identifier/keywords
			token = l.read_identifier()
		} else if is_digit(cur_char) {
			// handle _ error later
			token = l.read_number()
		} else {
			l.next_char()
			token = l.make_singlechar_token(ERROR, cur_char)
		}
	}

	return token
}

func (l *Lexer) make_token(tokentype TokenType, literal string) ProtoToken {
	return ProtoToken{
		Type:    tokentype,
		Literal: literal,
		TokenSpan: Span{
			Line:  l.line,
			Col:   l.column - 2,
			Start: l.pos - 2,
			End:   l.peek_pos,
		},
	}
}

func (l *Lexer) make_singlechar_token(tokentype TokenType, char byte) ProtoToken {
	return ProtoToken{
		Type:    tokentype,
		Literal: string(char),
		TokenSpan: Span{
			Line:  l.line,
			Col:   l.column - 1,
			Start: l.pos - 1,
			End:   l.peek_pos,
		},
	}
}

func (l *Lexer) read_string() ProtoToken {
	start := l.pos
	col := l.column
	l.next_char()
	for !l.is_at_end() && l.cur_byte != '"' {
		ch := l.cur_byte
		if ch == '\\' && l.peek_char() == '"' { // an escaped char
			l.next_char() // move to the ", will be skipped in the next_char() call below
		}
		if l.cur_byte == '\n' { // no multiline strings
			break
		}
		l.next_char()
	}

	if l.is_at_end() || l.cur_byte != '"' {
		return ProtoToken{
			Type:    ERROR,
			Literal: "Unterminated string literal",
			TokenSpan: Span{
				Line:  l.line,
				Col:   l.column - 1,
				Start: start,
				End:   l.pos,
			},
		}
	}
	l.next_char()
	slice := l.source[start:l.pos]
	return ProtoToken{
		Type:    STRING,
		Literal: slice,
		TokenSpan: Span{
			Line:  l.line,
			Col:   col,
			Start: start,
			End:   l.pos,
		},
	}
}

func (l *Lexer) read_character() ProtoToken {
	start := l.pos
	col := l.column
	l.next_char()

	if l.cur_byte == '\'' { //empty char
		l.next_char()
		return ProtoToken{
			Type:    CHAR,
			Literal: "",
			TokenSpan: Span{
				Line:  l.line,
				Col:   col,
				Start: start,
				End:   l.pos,
			},
		}
	} else { // contains some character
		char := l.cur_byte
		l.next_char()
		if l.cur_byte != '\'' { // character is either too long or unterminated
			return ProtoToken{
				Type:    ERROR,
				Literal: "Character literal unterminated (or contains more than 1 character)",
				TokenSpan: Span{
					Line:  l.line,
					Col:   col,
					Start: start,
					End:   l.pos,
				},
			}
		} else {
			l.next_char()
			return ProtoToken{
				Type:    CHAR,
				Literal: string(char),
				TokenSpan: Span{
					Line:  l.line,
					Col:   col,
					Start: start,
					End:   l.pos,
				},
			}
		}
	}
}

func (l *Lexer) read_identifier() ProtoToken {
	start := l.pos
	col := l.column
	l.next_char() // move to next character
	for !l.is_at_end() &&
		(is_alphabet(l.cur_byte) ||
			is_digit(l.cur_byte) ||
			l.cur_byte == '_') {
		l.next_char()
	}
	slice := l.source[start:l.pos]
	tokentype := CheckPotentialKeyword(slice)
	return ProtoToken{
		Type:    tokentype,
		Literal: slice,
		TokenSpan: Span{
			Line:  l.line,
			Col:   col,
			Start: start,
			End:   l.pos,
		},
	}
}

func (l *Lexer) read_number() ProtoToken {
	start := l.pos
	col := l.column
	l.next_char()
	for is_digit(l.cur_byte) {
		l.next_char()
	}
	slice := l.source[start:l.pos]

	// check if number is in i64 range
	if _, err := strconv.ParseInt(slice, 10, 64); err == nil {
		return ProtoToken{
			Type:    I64,
			Literal: slice,
			TokenSpan: Span{
				Line:  l.line,
				Col:   col,
				Start: start,
				End:   l.pos,
			},
		}
	} else {
		return ProtoToken{
			Type:    ERROR,
			Literal: slice + " isn't i64",
			TokenSpan: Span{
				Line:  l.line,
				Col:   col,
				Start: start,
				End:   l.pos,
			},
		}
	}
}

File: src/frontend/lexer.rs
use super::errors::LexError;
use super::source::SourceFile;
use super::token::Token;

#[allow(dead_code)]
pub struct Lexer {
    pub src: SourceFile,
    last_lexed: Option<Token>,
}

#[allow(dead_code)]
impl Lexer {
    pub fn new(src: SourceFile) -> Lexer {
        Lexer {
            src,
            last_lexed: None,
        }
    }

    // skip all whitespace characters
    fn skip_whitespace(&mut self) {
        loop {
            let c = self.src.cur_char();
            if c.is_whitespace() {
                self.src.next_char();
                continue;
            } else {
                break;
            }
        }
    }

    // process comments
    fn process_comment(&mut self) {
        let start = self.src.cur_char();
        let peek = self.src.peek_char();

        if start == '/' && peek == '/' {
            // consume both "//" characters
            self.src.next_char();
            self.src.next_char();

            // single line comment
            // skip until we reach a newline
            loop {
                let cur = self.src.cur_char();
                self.src.next_char();
                if cur != '\n' {
                    continue;
                } else {
                    break;
                }
            }
        }
    }

    // try to lex the next possible token
    pub fn next_token(&mut self) -> Result<Token, LexError> {
        while self.src.cur_char().is_whitespace()
            || (self.src.cur_char() == '/' && self.src.peek_char() == '/')
        {
            // skip all preceding whitespace
            self.skip_whitespace();
            // process comments
            // eventually, it'll get documentation comments
            // for now, it just skips single line comments
            self.process_comment();
        }

        // get the current character
        let c = self.src.cur_char();

        // if we reached the end of the file, return EOF
        if c == '\0' {
            return Ok(Token::Eof(self.src.get_ref()));
        }
        let maybe_token = match c {
            // operators
            '+' | '-' | '*' | '/' | '%' | '!' | '=' | '<' | '>' | '(' | ')' | '{' | '}' | '['
            | ']' | ',' | '.' | ':' | ';' | '^' | '$' | '@' => self.lex_operator(),
            // numbers
            _ if c.is_ascii_digit() => self.lex_number(),
            // identifiers | keywords
            _ if c.is_alphabetic() || c == '_' => self.lex_potential_identifier(),
            // ' which is used to start characters or strings
            '\'' => self.lex_char(),
            // " which is used to start strings
            '"' => self.lex_string(),
            // invalid character
            _ => {
                let mut err_ref = self.src.get_ref();
                err_ref.end_col += 1;
                let err = Err(LexError::InvalidCharacter(err_ref));
                self.src.next_char(); // try to keep lexing
                err
            }
        };

        match &maybe_token {
            Ok(t) => self.last_lexed = Some(t.clone()),
            Err(_) => self.last_lexed = None,
        }
        maybe_token
    }

    // lex a potential identifier
    // it might turn out to be either a:
    // - keyword
    // - identifier
    fn lex_potential_identifier(&mut self) -> Result<Token, LexError> {
        let mut id = String::new();
        let c_ref = self.src.get_ref();
        let cur = self.src.cur_char();
        id.push(cur);

        // read all characters that are alphanumeric or '_'
        let mut end_ref;
        loop {
            let c = self.src.next_char();
            end_ref = self.src.get_ref();
            if c.is_alphanumeric() || c == '_' {
                id.push(c);
            } else {
                break;
            }
        }

        let combined_ref = c_ref.combine(end_ref);
        // check if the identifier is a keyword
        match id.as_str() {
            "fn" => Ok(Token::Fn(combined_ref)),
            "let" => Ok(Token::Let(combined_ref)),
            "mut" => Ok(Token::Mut(combined_ref)),
            "if" => Ok(Token::If(combined_ref)),
            "else" => Ok(Token::Else(combined_ref)),
            "loop" => Ok(Token::Loop(combined_ref)),
            "while" => Ok(Token::While(combined_ref)),
            "void" => Ok(Token::Void(combined_ref)),
            "true" => Ok(Token::True(combined_ref)),
            "false" => Ok(Token::False(combined_ref)),
            "break" => Ok(Token::Break(combined_ref)),
            "continue" => Ok(Token::Continue(combined_ref)),
            "return" => Ok(Token::Return(combined_ref)),
            "use" => Ok(Token::Use(combined_ref)),
            "pub" => Ok(Token::Pub(combined_ref)),
            "mod" => Ok(Token::Mod(combined_ref)),
            "as" => Ok(Token::As(combined_ref)),

            "and" => Ok(Token::And(combined_ref)),
            "or" => Ok(Token::Or(combined_ref)),
            "not" => Ok(Token::Not(combined_ref)),

            "i8" => Ok(Token::I8(combined_ref)),
            "i16" => Ok(Token::I16(combined_ref)),
            "i32" => Ok(Token::I32(combined_ref)),
            "i64" => Ok(Token::I64(combined_ref)),
            "isize" => Ok(Token::Isize(combined_ref)),
            "u8" => Ok(Token::U8(combined_ref)),
            "u16" => Ok(Token::U16(combined_ref)),
            "u32" => Ok(Token::U32(combined_ref)),
            "u64" => Ok(Token::U64(combined_ref)),
            "usize" => Ok(Token::Usize(combined_ref)),
            "bool" => Ok(Token::Bool(combined_ref)),
            "char" => Ok(Token::Char(combined_ref)),
            "str" => Ok(Token::Str(combined_ref)),
            _ => Ok(Token::Identifier(id, combined_ref)),
        }
    }

    // lex an operator
    fn lex_operator(&mut self) -> Result<Token, LexError> {
        let cur_ref = self.src.get_ref();
        let cur = self.src.cur_char();
        // detect a signed number
        if cur == '-' {
            // if the next character is a digit, lex a signed number
            let peek = self.src.peek_char();
            if peek.is_ascii_digit() {
                return self.lex_signed_number();
            } else {
                // otherwise, return a minus operator
                self.src.next_char();
                return Ok(Token::Minus(cur_ref.combine(self.src.get_ref())));
            }
        }

        match cur {
            // arithmetic operators
            '+' => {
                self.src.next_char();
                Ok(Token::Plus(cur_ref.combine(self.src.get_ref())))
            }
            '-' => {
                self.src.next_char();
                Ok(Token::Minus(cur_ref.combine(self.src.get_ref())))
            }
            '*' => {
                self.src.next_char();
                Ok(Token::Star(cur_ref.combine(self.src.get_ref())))
            }
            '/' => {
                self.src.next_char();
                Ok(Token::Slash(cur_ref.combine(self.src.get_ref())))
            }
            '%' => {
                self.src.next_char();
                Ok(Token::Modulo(cur_ref.combine(self.src.get_ref())))
            }

            // special characters
            '^' => {
                self.src.next_char();
                Ok(Token::Caret(cur_ref.combine(self.src.get_ref())))
            }
            '@' => {
                self.src.next_char();
                Ok(Token::At(cur_ref.combine(self.src.get_ref())))
            }
            '$' => {
                self.src.next_char();
                Ok(Token::Dollar(cur_ref.combine(self.src.get_ref())))
            }

            // logical operators
            '!' => {
                // if the next character is a '=', return a NotEqual operator
                let c = self.src.peek_char();
                if c == '=' {
                    self.src.next_char();
                    self.src.next_char();
                    return Ok(Token::NotEqual(cur_ref.combine(self.src.get_ref())));
                }
                // otherwise, return a Not operator
                self.src.next_char();
                Ok(Token::Exclamation(cur_ref.combine(self.src.get_ref())))
            }
            '=' => {
                // if the next character is a '=', return a Equal operator
                let c = self.src.peek_char();
                if c == '=' {
                    self.src.next_char();
                    self.src.next_char();
                    return Ok(Token::Equal(cur_ref.combine(self.src.get_ref())));
                }
                // otherwise, return a Assign operator
                self.src.next_char();
                Ok(Token::Assign(cur_ref.combine(self.src.get_ref())))
            }
            '<' => {
                // if the next character is a '=', return a LessEqual operator
                let c = self.src.peek_char();
                if c == '=' {
                    self.src.next_char();
                    self.src.next_char();
                    return Ok(Token::LessEqual(cur_ref.combine(self.src.get_ref())));
                }
                // otherwise, return a Less operator
                self.src.next_char();
                Ok(Token::Less(cur_ref.combine(self.src.get_ref())))
            }
            '>' => {
                // if the next character is a '=', return a GreaterEqual operator
                let c = self.src.peek_char();
                if c == '=' {
                    self.src.next_char();
                    self.src.next_char();
                    return Ok(Token::GreaterEqual(cur_ref.combine(self.src.get_ref())));
                }
                // otherwise, return a Greater operator
                self.src.next_char();
                Ok(Token::Greater(cur_ref.combine(self.src.get_ref())))
            }
            ':' => {
                // if the next character is a ':', return a Scope operator
                let c = self.src.peek_char();
                if c == ':' {
                    self.src.next_char();
                    self.src.next_char();
                    return Ok(Token::Scope(cur_ref.combine(self.src.get_ref())));
                }
                // otherwise, return a Colon operator
                self.src.next_char();
                Ok(Token::Colon(cur_ref.combine(self.src.get_ref())))
            }
            '(' => {
                self.src.next_char();
                Ok(Token::LParen(cur_ref.combine(self.src.get_ref())))
            }
            ')' => {
                self.src.next_char();
                Ok(Token::RParen(cur_ref.combine(self.src.get_ref())))
            }
            '{' => {
                self.src.next_char();
                Ok(Token::LCurly(cur_ref.combine(self.src.get_ref())))
            }
            '}' => {
                self.src.next_char();
                Ok(Token::RCurly(cur_ref.combine(self.src.get_ref())))
            }
            '[' => {
                self.src.next_char();
                Ok(Token::LBracket(cur_ref.combine(self.src.get_ref())))
            }
            ']' => {
                self.src.next_char();
                Ok(Token::RBracket(cur_ref.combine(self.src.get_ref())))
            }
            ',' => {
                self.src.next_char();
                Ok(Token::Comma(cur_ref.combine(self.src.get_ref())))
            }
            '.' => {
                self.src.next_char();
                Ok(Token::Dot(cur_ref.combine(self.src.get_ref())))
            }
            ';' => {
                self.src.next_char();
                Ok(Token::Semicolon(cur_ref.combine(self.src.get_ref())))
            }
            _ => unreachable!("invalid operator character: '{}' at {:?}", cur, cur_ref),
        }
    }

    fn lex_string(&mut self) -> Result<Token, LexError> {
        let mut span = self.src.get_ref();
        let mut content = String::new();

        // read all characters until terminating "
        while !self.src.is_eof() {
            let c = self.src.next_char();
            span = span.combine(self.src.get_ref());
            if c == '"' {
                break;
            } else if c == '\\' {
                // if the next character is a ", add it to the content
                let c = self.src.next_char();
                span = span.combine(self.src.get_ref());
                if c == '"' {
                    content.push('"');
                } else {
                    // otherwise, add the escape character
                    content.push('\\');
                    content.push(c);
                }
            } else {
                content.push(c);
            }
        }

        // if the string is not terminated, return an error
        if self.src.is_eof() {
            Err(LexError::UnterminatedStringLiteral(span))
        } else {
            self.src.next_char();
            span = span.combine(self.src.get_ref());
            Ok(Token::StringLiteral(span, content))
        }
    }

    // try to lex a character
    fn lex_char(&mut self) -> Result<Token, LexError> {
        let mut span = self.src.get_ref();
        let mut content = String::new();

        // read single character
        let c = self.src.next_char();
        span = span.combine(self.src.get_ref());

        // if the next character is a ', return an error
        if c == '\'' {
            return Err(LexError::EmptyCharacterLiteral(span));
        } else if c == '\\' {
            // if the next character is a ', add it to the content
            let c = self.src.next_char();
            span = span.combine(self.src.get_ref());
            if c == '\'' {
                content.push('\'');
            } else {
                // otherwise, add the escape character
                content.push('\\');
                content.push(c);
            }
        } else {
            content.push(c);
        }

        // if the next character is not a ', return an error
        let c = self.src.next_char();
        span = span.combine(self.src.get_ref());
        if c != '\'' {
            return Err(LexError::UnterminatedCharacterLiteral(span));
        }

        self.src.next_char();
        span = span.combine(self.src.get_ref());
        Ok(Token::CharLiteral(span, content.chars().next().unwrap()))
    }

    // try to lex a number
    fn lex_number(&mut self) -> Result<Token, LexError> {
        // by default, calling lex_number will lex a signed number
        // since users tend to perform signed arithmetic more often
        self.lex_signed_number()
    }

    // try to lex an unsigned number which is one of:
    // - i8
    // - i16
    // - i32
    // - i64
    // - isize
    fn lex_signed_number(&mut self) -> Result<Token, LexError> {
        let mut signed_number = String::new();
        let cur_ref = self.src.get_ref();
        let cur = self.src.cur_char();
        signed_number.push(cur);

        // read all characters that are digits
        let mut end_ref;
        loop {
            let c = self.src.next_char();
            end_ref = self.src.get_ref();
            if c.is_ascii_digit() {
                signed_number.push(c);
            } else if c == '_' {
                // ignore underscores
                continue;
            } else {
                break;
            }
        }

        let combined_ref = cur_ref.combine(end_ref);
        // check if signed_number is i8
        if let Ok(num) = signed_number.parse::<i8>() {
            return Ok(Token::I8Literal(num, combined_ref));
        }

        if let Ok(num) = signed_number.parse::<i16>() {
            return Ok(Token::I16Literal(num, combined_ref));
        }

        if let Ok(num) = signed_number.parse::<i32>() {
            return Ok(Token::I32Literal(num, combined_ref));
        }

        if let Ok(num) = signed_number.parse::<i64>() {
            return Ok(Token::I64Literal(num, combined_ref));
        }

        if let Ok(num) = signed_number.parse::<isize>() {
            return Ok(Token::IsizeLiteral(num, combined_ref));
        }

        // if signed_number's first character is not '-',
        // then attempt to lex an unsigned number
        self.src.jump_to(&cur_ref);
        self.lex_unsigned_number()
    }

    // try to lex an unsigned number which is one of:
    // - u8
    // - u16
    // - u32
    // - u64
    // - usize
    fn lex_unsigned_number(&mut self) -> Result<Token, LexError> {
        let mut unsigned_number = String::new();
        let cur_ref = self.src.get_ref();
        let cur = self.src.cur_char();
        unsigned_number.push(cur);

        // read all characters that are digits
        let mut end_ref;
        loop {
            let c = self.src.next_char();
            end_ref = self.src.get_ref();
            if c.is_ascii_digit() {
                unsigned_number.push(c);
            } else if c == '_' {
                // ignore underscores
                continue;
            } else {
                break;
            }
        }

        let combined_ref = cur_ref.combine(end_ref);
        // check if unsigned_number is u8
        if let Ok(num) = unsigned_number.parse::<u8>() {
            return Ok(Token::U8Literal(num, combined_ref));
        }

        if let Ok(num) = unsigned_number.parse::<u16>() {
            return Ok(Token::U16Literal(num, combined_ref));
        }

        if let Ok(num) = unsigned_number.parse::<u32>() {
            return Ok(Token::U32Literal(num, combined_ref));
        }

        if let Ok(num) = unsigned_number.parse::<u64>() {
            return Ok(Token::U64Literal(num, combined_ref));
        }

        if let Ok(num) = unsigned_number.parse::<usize>() {
            return Ok(Token::UsizeLiteral(num, combined_ref));
        }

        // if it is not a valid unsigned number, return an error
        Err(LexError::CannotMakeUnsignedNumber(combined_ref))
    }
}

#[allow(dead_code)]
#[derive(serde::Deserialize, serde::Serialize, Debug)]
struct LexerTestResult {
    stringified_tokens: Vec<String>,
}

#[test]
fn test_lexer() {
    insta::glob!("lexer_inputs/*.pr", |path| {
        // build the SourceFile from the proto file
        let path = path.to_str().unwrap().to_string();
        let src: SourceFile = SourceFile::new(path);

        // build the lexer
        let mut lexer = Lexer::new(src);

        let mut tokens = Vec::new();
        let mut res = LexerTestResult {
            stringified_tokens: Vec::new(),
        };
        loop {
            let token = lexer.next_token();
            if token.is_err() {
                panic!("error lexing: {token:?}");
            }
            let token = token.unwrap();
            if let Token::Eof(_) = token {
                break;
            }
            res.stringified_tokens.push(format!(
                "{} at {}",
                token.as_str(),
                token.get_source_ref().as_str()
            ));
            tokens.push(token);
        }
        insta::assert_yaml_snapshot!(res)
    });
}
